### ISSUE TITLE
matterbridge with z2m - issues with device recognition and matter descriptors.

### ISSUE THREAD

**Issue by jpadie** [Luligu/matterbridge#17]
### Discussed in https://github.com/Luligu/matterbridge/discussions/16

<div type='discussions-op-text'>

<sup>Originally posted by **jpadie** April 26, 2024</sup>
z2m configured and working fine for ages.
matterbridge installs ok.  idem the plugin.
plugin configured with mqtt details.  
however in the matterbridge front end the z2m plugin is showing zero devices, although two are registered on the z2m front end.  messages come in via mqtt and are shown in the logs as warnings.  

took a few minutes for the devices in z2m to be found by matterbridge.  in the meantime mqtt messages showed in the logs as warning. 

the two devices on the coordinator are a sonoff temp and humidity sensor and a 2-gang dimmer module.

the temp and humidity module correctly shows its data on the matterbridge front end. but alexa only sees the temperature.  despite alexa docs there is a good device ui available for atmospheric information sensors since this module can be paired directly to alexa and is represented on a reasonable gui rather than as two separate devices. 

the coordinator appears to be shown in alexa and in the matterbridge front end as a door lock. clearly this is not correct.  there is no reason for the coordinator to be published in alexa at all, imo.

the 2 gang dimmer module is exposed to alexa as two lights; one responsive device that has no dimmer capability and one device that is unresponsive. matterbridge describes the module as a "bridgedevice-0x00013" and shows no useful information in the cluster column. both gangs work perfectly with Z2M.  this is a TuYa device, by the way.

so there are two devices in Z2M. matterbridge records 3 and reports to alexa that there are 5.  that probably makes sense if the matterbridge is reporting the temp and humidity sensor as two separate clusters rather than combined (alexa does support combined devices now, from memory).

happy to provide more material if that is helpful in debugging. will try to convert this to an issue.</div>

**Comment by Luligu** [comment]
Hi,
thank you for your interest and your message.
I need some more information...
First I need the full log of Matterbridge.
Then in the Matterbridge directory (root of your home directory) there is matterbridge-zigbee2mqtt directory. Inside you find 3 files bridge... I need those 3 files.
The Coordinator is exposed like a DoorLock because this allows you to change permit-join asking alexa in your case to open the Coordinator. If you don't like it you can simply put the coordinator in the BlackList: "blackList": ["Coordinator"], 
After the number 0 in the GUI means that Matterbridge is waiting for the controler to connect so is not a problem. You can update the page.
In the Devices you can have an idea of what is passed to Alexa and see is something seems wrong...
With the files I asked you I can be more precise.

**Comment by jpadie** [comment]
thank you for responding and the explanation about the door-lock.  that makes sense now. 

the files you requested are attached.  i have redacted part of bridge-info.json for security.  the same data is redacted from the logs (also attached).
[bridge-devices.json](https://github.com/Luligu/matterbridge/files/15135742/bridge-devices.json)
[matterbridge.log](https://github.com/Luligu/matterbridge/files/15135743/matterbridge.log)
[bridge-info.json](https://github.com/Luligu/matterbridge/files/15135744/bridge-info.json)
[bridge-groups.json](https://github.com/Luligu/matterbridge/files/15135745/bridge-groups.json)

**Comment by jpadie** [comment]
The second light is now responsive.  Not sure when it came on line.  However neither can be dimmed.

One of the five devices that were published to Alexa was the composed device of the two lights.  So the humidity sensor was not reported to Alexa at all. Alexa does support air quality monitor endpoints and rangecontroller endpoints which could be used for humidity sensors (if there is no direct support - I've certainly had humidity sensors in my dev Alexa device list in the past)

**Comment by Luligu** [comment]
The Message for ***unknown*** entity like warning are not a problem for Matterbridge. It's just your broker that has still retained messages from removed devices. You can reset the persistence if you want.
All your devices are totally supported. Nothing is wrong with the definition. 
In a discussion a user managed to use Alexa without problem.

**Comment by jpadie** [comment]
I disagree that nothing is wrong.No dimmer function for each of the lights and no humidity device reported to Alexa.  These appear to be issues, do they not?On 27 Apr 2024 20:52, Luligu ***@***.***> wrote:
The Message for unknown entity like warning are not a problem for Matterbridge. It's just your broker that has still retained messages from removed devices. You can reset the persistence if you want.
All your devices are totally supported. Nothing is wrong with the definition.
In a discussion a user managed to use Alexa without problem.

**Comment by jpadie** [comment]
No dimmer function for each of the lights and no humidity device reported to Alexa.  

These appear to be issues, do they not?

**Comment by tammeryousef1006** [comment]
i did manage to connect the bridge to several controllers , google home , home assistant , smartthings , you can say that home assistant was the only controller that shown all supported functions , when it comes to other controllers each one manage in different way , google home paired correctly and shown the devices offline and after reboot the google home hub all devices gets online with accepted functions , smartthings paired correctly but without control , after reboot the hub all devices started to work correctly , some devices with multi features worked normally , other devices shown one function only etc .

**Comment by jpadie** [comment]
I

> i did manage to connect the bridge to several controllers , google home , home assistant , smartthings , you can say that home assistant was the only controller that shown all supported functions , when it comes to other controllers each one manage in different way , google home paired correctly and shown the devices offline and after reboot the google home hub all devices gets online with accepted functions , smartthings paired correctly but without control , after reboot the hub all devices started to work correctly , some devices with multi features worked normally , other devices shown one function only etc .

I am certain I have paired virtual matter dimmers with alexa before.  I've just tested with a tasmotised esp32 and with chip-tool and both create matter dimmers in Alexa without problems.  I can try with matter.js too (at least an old version that I have installed) if that is a useful test.

**Comment by tammeryousef1006** [comment]
> I
> 
> > i did manage to connect the bridge to several controllers , google home , home assistant , smartthings , you can say that home assistant was the only controller that shown all supported functions , when it comes to other controllers each one manage in different way , google home paired correctly and shown the devices offline and after reboot the google home hub all devices gets online with accepted functions , smartthings paired correctly but without control , after reboot the hub all devices started to work correctly , some devices with multi features worked normally , other devices shown one function only etc .
> 
> I am certain I have paired virtual matter dimmers with alexa before. I've just tested with a tasmotised esp32 and with chip-tool and both create matter dimmers in Alexa without problems. I can try with matter.js too (at least an old version that I have installed) if that is a useful test.


its different issue , you are talking of pairing a device , but your problem with pairing a bridge controller , i will try to get alexa hub and check the results soon and give feedback about it

**Comment by jpadie** [comment]
Can you clarify?  
Tasmota devices commission as an aggregator.  Then each endpoint is automatically deployed to matter controllers.  Per standard 

So a dimmer is seen by Alexa as an aggregator with one other endpoint.   Three dimmers = one aggregator + 3 other endpoints and so on.  I have not tried a composed device as I don't think that is possible in tasmota.   I can try with chip-tool.  But a composed device is not a necessary solution as the bridge can disaggregate a 2 gang device into two separate single endpoints.

**Comment by tammeryousef1006** [comment]
let me check it once i got an alexa hub sooner , i will test and get back to you

**Comment by jpadie** [comment]
@Luligu  - Why is this closed as "completed"?  Have you pushed a fix for these issues?

**Comment by Luligu** [comment]
Was just 1 month old...
Are you still having the issue?
With Alexa there should be no issue at all.
The next release will change the way the bridge load... 
Still under test...

**Comment by Luligu** [comment]
If you have another issue open a new issue please.
This issue is actually too old and the latest is totally different now.

**Comment by tammeryousef1006** [comment]
> @Luligu - Why is this closed as "completed"? Have you pushed a fix for these issues?

latest update solve all issues with google home and alexa , I pair with alexa hub since the first try and all goes smooth 

![fab new](https://github.com/Luligu/matterbridge/assets/62266828/d0393dd4-840a-4fb0-b980-b3b54d62a23e)

**Comment by jpadie** [comment]
Demonstrably and evidently this is not the case.  Otherwise I would not have reported the continued existence of the bug.On 8 Jun 2024 22:22, Tamer Salah ***@***.***> wrote:

@Luligu - Why is this closed as "completed"? Have you pushed a fix for these issues?

latest update solve all issues with google home and alexa , I pair with alexa hub since the first try and all goes smooth
fab.new.png (view on web)

**Comment by tammeryousef1006** [comment]
![Screenshot_20240609_013712_Home](https://github.com/Luligu/matterbridge/assets/62266828/d1c94fcb-15b5-489f-904c-2fa7b22517c5)
![Screenshot_20240609_013640_SmartThings](https://github.com/Luligu/matterbridge/assets/62266828/3f8ba6a2-a1cb-4cdd-8eb3-b30869d6e09d)
![Screenshot_20240609_013617_Amazon Alexa](https://github.com/Luligu/matterbridge/assets/62266828/f4e1e1bb-fd1a-4ead-9d3c-09151dc377a5)

Dear @Luligu  kindly take a look at this device for example which has multiple functions but controllers behave in different way

**Comment by tammeryousef1006** [comment]
after you update the matterbridge and zigbee2mqtt plugin , which issues still remaining  ?

**Comment by jpadie** [comment]
The lights are not configured by Alexa as having brightness control.   They work as expected in z2m. I have paired brightness control and colour capable lights via matter with Alexa so I know this is not an Alexa support issue.  Ditto combined temperature/ humidity sensors.  The temperature is passed to Alexa but the humidity is not.  Supposedly Alexa does not yet officially support humidity sensors via matter (and other pathways) but I have paired humidity sensor devices via the API and directly via ZigBee with success so I am aware that they are "somewhat" supported however the workaround may be to present the values as some other percentage and add a user label. On 9 Jun 2024 00:07, Tamer Salah ***@***.***> wrote:
after you update the matterbridge and zigbee2mqtt plugin , which issues still remaining  ?

**Comment by Luligu** [comment]
![Screenshot_20240610_203211_Amazon_Alexa](https://github.com/Luligu/matterbridge/assets/132135057/e2704037-52bc-4df2-9830-5ddfa19f167a)
The humidity sensor has no issue as you see. 
Alexa in test done doesn't show pressure and flow sensor device type. It also doesn't show the light sensor device type.

**Comment by jpadie** [comment]
However the issue remains in your code, unless you have fixed the issue in your code base since the issue was reported.  In which case please link to the release and change log that shows the fix.

**Comment by tammeryousef1006** [comment]
as tests before with google home , smartthings & alexa all the tested devices were running fine and stable , however if there is such a device with multi gang like yours it will be easier to study if we had it to be able to see the output at log and commands behavior , but anyway alexa was the only controller was showing less functionality than smartthings and google home , so if you want support for multi gang dimmer switch kindly upload logs after rebooting mb and sending commands from zigbee2mqtt and from alexa so we can take a deeper look to detect any issue or errors in order to adjust or tweak the plugin , thanks for follow up in advance

**Comment by jpadie** [comment]
I can supply logs of course. However the issue is not in the z2m layer (thus logs of the device talking will not be helpful), but in the way that matterbridge composes devices, which appears to be in a way that is not expressly provided for in the matter specs (multiple devices inside an endpoint that is not a composed endpoint or an aggregator).   As expressed several times, for compliance reasons it would be better to expose each device as a single child of the aggregator or as a separate child of a composed device (bridgedEndpoint). 

However much more valuable is to supply example code based on matter.js that actually works and shows the way that composed devices _should_ be promoted to controllers.  This I have done.

For devices, I have found that all moes and other tuya modules with more than one gang that I have teated behave the same way.  I would be happy to have one sent to you if you pm me your address (assuming you are in a country with good support from amazon).

**Comment by tammeryousef1006** [comment]
i got a new moes 3 gang switch but without dimmer , i will connect it today and give it a test and see if it works

**Comment by jpadie** [comment]
My experience is that matterbridge promotes them to Alexa as onoff devices.  They show up  ok and can be turned on/off.  

Where does matterbridge log the endpoint structure that it promotes to a controller?  I'm looking for the identical format to logendpoint in matter.js?

**Comment by Luligu** [comment]
> I can supply logs of course. However the issue is not in the z2m layer (thus logs of the device talking will not be helpful), but in the way that matterbridge composes devices, which appears to be in a way that is not expressly provided for in the matter specs (multiple devices inside an endpoint that is not a composed endpoint or an aggregator). As expressed several times, for compliance reasons it would be better to expose each device as a single child of the aggregator or as a separate child of a composed device (bridgedEndpoint).

That is exactly what we do: child endpoints of a bridgedEndpoint (device type 0x0013).

**Comment by Luligu** [comment]
> My experience is that matterbridge promotes them to Alexa as onoff devices. They show up ok and can be turned on/off.
> 
> Where does matterbridge log the endpoint structure that it promotes to a controller? I'm looking for the identical format to logendpoint in matter.js?


Matterbridge has a frontend, Devices section...

**Comment by jpadie** [comment]
Then the question is "what happens when you use my code snip”? 

If you do not get what I do then your codebase is doing something different.  Very simple test and a proven method of obtaining a working solution.

**Comment by jpadie** [comment]
i have added endpoint logging to your code and retrieved the logs from syslog.   I advise you to post in your readme how to do this and how to tell the rsyslog daemon not to escape your colour coding of the logs; or to follow the more usual mechanism of not colour coding log entries.  Either way, the syslog (or other system level log) is a much better place to debug from than a console that may not appear in time to capture the information required.  

**turning first to the issue of composed devices**

I can see in the snip attached (line 351 - 483) that you are not using a bridged endpoint (0x0013) to hold the temp sensor and humidity sensor.  instead you are combining clusters in one endpoint.  

the matter specification **does not define** how a controller should handle a device type with multiple clusters that are not mandated or specified as optional.  Thus it follows that it is not currently within specification to combine such clusters in one device type.

Instead the matter specs define a bridged node (0x0013) intended to hold foreign network devices.   Until the matter specifications specifically permit arbitrary clusters added to a device would it not be prudent to use 0x0013 as the parent of multiple clusters?  

I also note that you include a power source configuration cluster in the endpoint definition for a temperature sensor.  this does not appear to be supported by the matter specification in the same way that composed devices are only explicitly supported inside a bridge node.  however the cluster is supported for a bridged node (0x0013), which makes more sense.  

In this specific case, in v1.2 there is a device type available for a temp + RH sensor.  however until 1.2 or later has wide adoption it seems sensible to use the spec-compliant mechanism of building the device within a bridge.

interestingly in the schema that you capture from z2m a bridge node is specified in the json for this physical device. 

**Now let's look at the bridged endpoint**.  

lines 484-.  This endpoint contains the two lights that z2m correctly identifies as dimmable.  

In the first light you can see that your code has not included the level control cluster (0x0008).  or more properly identified the device as a DimmableLightDevice (0x0101).  the same is true of the second light.

in the logs it appears that your code does correctly recognise the input from z2m as being dimmable lights with the labels l1 and l2. 
```

12:53:28.641] [Matterbridge zigbee2mqtt plugin] Registering device 0x00124b002a5095ef ID: 0x00124b002a5095ef 
12:53:28.656] [0x00124b002a5095ef] Created MatterEntity: 0x00124b002a5095ef 
12:53:28.680] [0x00124b002a5095ef] Device 0x00124b002a5095ef endpoint:  type:  property: temperature => deviceT>
12:53:28.733] [0x00124b002a5095ef] new BridgedBaseDevice 0x00124b002a5095ef deviceType: MA-tempsensor 
12:53:28.745] [0x00124b002a5095ef] - with cluster: 1026-TemperatureMeasurement 
12:53:28.756] [0x00124b002a5095ef] - with cluster: 3-Identify 
12:53:28.775] [0x00124b002a5095ef] - with cluster: 1026-TemperatureMeasurement 
12:53:28.901] [0x00124b002a5095ef] Device 0x00124b002a5095ef endpoint:  type:  property: humidity => deviceType>
12:53:28.907] [0x00124b002a5095ef] addDeviceTypeAndClusterServer deviceType: MA-humiditysensor 
12:53:28.924] [0x00124b002a5095ef] - with cluster: 3-Identify 
12:53:28.936] [0x00124b002a5095ef] - with cluster: 1029-RelativeHumidityMeasurement 
12:53:28.940] [0x00124b002a5095ef] - with cluster: 1029-RelativeHumidityMeasurement 
12:53:28.962] [Matterbridge] Adding bridged device 0x00124b002a5095ef (MA-tempsensor) for plugin matterbridge-z>
12:53:28.986] [Matterbridge] Added and registered bridged device (2/2) 0x00124b002a5095ef (MA-tempsensor) for p>
12:53:28.993] [Matterbridge zigbee2mqtt plugin] Registered device 0x00124b002a5095ef ID: 0x00124b002a5095ef 
12:53:29.002] [Matterbridge zigbee2mqtt plugin] Registering device 0x187a3efffef17704 ID: 0x187a3efffef17704 
12:53:29.021] [0x187a3efffef17704] Created MatterEntity: 0x187a3efffef17704 
12:53:29.036] [0x187a3efffef17704] Device 0x187a3efffef17704 endpoint: l1 type: light property: state => device>
12:53:29.064] [0x187a3efffef17704] new BridgedBaseDevice 0x187a3efffef17704 deviceType: MA-bridgeddevice 
12:53:29.174] [0x187a3efffef17704] addChildDeviceTypeAndClusterServer: Child endpoint created: l1 
12:53:29.246] [0x187a3efffef17704] - with cluster: 3-Identify 
12:53:29.259] [0x187a3efffef17704] - with cluster: 4-Groups 
12:53:29.274] [0x187a3efffef17704] - with cluster: 6-OnOff 
12:53:29.284] [0x187a3efffef17704] - with cluster: 6-OnOff 
12:53:29.364] [0x187a3efffef17704] Device 0x187a3efffef17704 endpoint: l1 type: light property: brightness => d>
12:53:29.379] [0x187a3efffef17704] addChildDeviceTypeAndClusterServer: Child endpoint found: l1 
12:53:29.385] [0x187a3efffef17704] - with cluster: 3-Identify 
12:53:29.401] [0x187a3efffef17704] - with cluster: 4-Groups 
12:53:29.417] [0x187a3efffef17704] - with cluster: 6-OnOff 
12:53:29.437] [0x187a3efffef17704] - with cluster: 8-LevelControl 
12:53:29.442] [0x187a3efffef17704] - with cluster: 8-LevelControl 
12:53:29.526] [0x187a3efffef17704] Device 0x187a3efffef17704 endpoint: l2 type: light property: state => device>
12:53:29.538] [0x187a3efffef17704] addChildDeviceTypeAndClusterServer: Child endpoint created: l2 
12:53:29.564] [0x187a3efffef17704] - with cluster: 3-Identify 
12:53:29.578] [0x187a3efffef17704] - with cluster: 4-Groups 
12:53:29.595] [0x187a3efffef17704] - with cluster: 6-OnOff 
12:53:29.604] [0x187a3efffef17704] - with cluster: 6-OnOff 
12:53:29.656] [0x187a3efffef17704] Device 0x187a3efffef17704 endpoint: l2 type: light property: brightness => d>
12:53:29.667] [0x187a3efffef17704] addChildDeviceTypeAndClusterServer: Child endpoint found: l2 
12:53:29.678] [0x187a3efffef17704] - with cluster: 3-Identify 
12:53:29.684] [0x187a3efffef17704] - with cluster: 4-Groups 
12:53:29.692] [0x187a3efffef17704] - with cluster: 6-OnOff 
12:53:29.711] [0x187a3efffef17704] - with cluster: 8-LevelControl 
12:53:29.724] [0x187a3efffef17704] - with cluster: 8-LevelControl 
12:53:29.763] [Matterbridge] Adding bridged device 0x187a3efffef17704 (MA-bridgeddevice) for plugin matterbridg>
12:53:29.776] [Matterbridge] Added and registered bridged device (3/3) 0x187a3efffef17704 (MA-bridgeddevice) fo>
12:53:29.790] [Matterbridge zigbee2mqtt plugin] Registered device 0x187a3efffef17704 ID: 0x187a3efffef17704 
```
in case useful, later in the logs the two lights are flagged as having no unique ID
```

Jun 12 12:53:30 raspberrypi matterbridge[27503]: 2024-06-12 12:53:30.194 DEBUG  CommissioningServer  Restored endpoint id 1 for endpoint with unique_CSb6ee40dc02>
Jun 12 12:53:30 raspberrypi matterbridge[27503]: 2024-06-12 12:53:30.212 DEBUG  CommissioningServer  Restored endpoint id 2 for endpoint with unique_CSb6ee40dc02>
Jun 12 12:53:30 raspberrypi matterbridge[27503]: 2024-06-12 12:53:30.239 DEBUG  CommissioningServer  Restored endpoint id 3 for endpoint with unique_CSb6ee40dc02>
Jun 12 12:53:30 raspberrypi matterbridge[27503]: 2024-06-12 12:53:30.256 DEBUG  CommissioningServer  Restored endpoint id 4 for endpoint with unique_CSb6ee40dc02>
Jun 12 12:53:30 raspberrypi matterbridge[27503]: 2024-06-12 12:53:30.264 DEBUG  CommissioningServer  No unique id found for endpoint on index 0 / device MA-onoff>
Jun 12 12:53:30 raspberrypi matterbridge[27503]: 2024-06-12 12:53:30.274 DEBUG  CommissioningServer  Restored endpoint id 5 for endpoint with unique_CSb6ee40dc02>
Jun 12 12:53:30 raspberrypi matterbridge[27503]: 2024-06-12 12:53:30.278 DEBUG  CommissioningServer  No unique id found for endpoint on index 1 / device MA-onoff>
Jun 12 12:53:30 raspberrypi matterbridge[27503]: 2024-06-12 12:53:30.281 DEBUG  CommissioningServer  Restored endpoint id 6 for endpoint with unique_CSb6ee40dc02>
Jun 12 12:53:30 raspberrypi matterbridge[27503]: 2024-06-12 12:53:30.522 DEBUG  CommissioningServer  Restoring attribute totalOperationalHours (3) in cluster Gen>
Jun 12 12:53:30 raspberrypi matterbridge[27503]: 2024-06-12 12:53:30.911 DEBUG  UdpChannelNode       Socket created and bound remoteAddress: undefined:5540 local>
Jun 12 12:53:30 raspberrypi matterbridge[27503]: 2024-06-12 12:53:30.923 DEBUG  UdpChannelNode       Socket created and bound remoteAddress: undefined:5540 local>

```

so there are two distinct issues that lead to this use case being a set of unresolved bugs:

1. you are relying on unspecified behaviour in the hopes that controllers will correctly handle devices with clusters that are not specified as optional or mandatory.  This leads to the Alexa ecosystem recognising only the first endpoint.  The Alexa behaviour **is** compliant with v1.0 specs (and in this case, as it happens, 1.3 specs).  

2. somewhere in your code it appears that you are not adding the level control cluster for lights that you have previously identified (from z2m) as dimmable.  

if it helps you chase down the bug in 2, these lines are frequently repeated in the logs (as posted yesterday but cleaned of your colour coding).  

 ```
2087	Jun 12 12:59:29 raspberrypi matterbridge[27503]: [12:59:29.083] [0x187a3efffef17704] Update endpoint 5 (l1) cluster 8-LevelControl not found: is z2m converter exposing all features? 
2088	Jun 12 12:59:29 raspberrypi matterbridge[27503]: [12:59:29.098] [0x187a3efffef17704] Update endpoint 6 (l2) cluster 8-LevelControl not found: is z2m converter exposing all features? 
```

As further help, in the entity.js file within the z2m plugin code, within the addChildDeviceTypeAndClusterServer method of the BridgedBaseDevice class there is a list of conditionals that add in the appropriate clusters.  I can see that there is no conditional for adding in either colour control  or level control.  I am not familiar enough with your code to determine whether this is the source of the bug articulated in number 2 but it does seem to be an oversight and indeed adding these lines at about line 1054 of entity.js seems to make things work as expected in Alexa.  I would have thought the same bug would have been present in all controller implementations.  I am surprised that you found it to be working in your testing.

```
 if (includeServerList.includes(LevelControl.Cluster.id)){
            child.addClusterServer(this.getDefaultLevelControlClusterServer());
        }
        
        if (includeServerList.includes(ColorControl.Cluster.id)){
            child.addClusterServer(this.getDefaultColorControlServer());
        }

```

Regrettably this is as much time as I can spend at the moment trying to help fix you bugs.  I trust you will, perhaps find it of some help.  

[matterbridge endpoint structure.txt](https://github.com/user-attachments/files/15805347/matterbridge.endpoint.structure.txt)

**Comment by Luligu** [comment]
1) log part
For the log the correct command is reported in the readme... 

View the log of Matterbridge in real time (this will show the log with colors)
sudo journalctl -u matterbridge.service -f --output cat

With that you go through the log with colors: then if you want, use it, if not let it go.

2) turning first to the issue of composed devices
all this paragraph has no sense to me:
"deviceTypeList" (0x0): value = [{"deviceType":770,"revision":2},{"deviceType":19,"revision":1},{"deviceType":775,"revision":2}]
these are exactly 770=0x302 775=0x307 19=0x13 alias MA-humiditysensor, MA-tempsensor and MA-bridgednode
so exactly what you expected to see!
PowerSource and PowerSourceConfiguration (now deprecated in Matter 1.3) can be with the device type 0x13 or with with the device type 0x11. In this case since there is 0x13 I didn't add 0x11. Both possibilities are well and clearly explained in Matter specs.
All other cluster are required by these 3 device types. Certainly not "arbitrary clusters added to a device".
I imagine you don't know the difference between a decimal number and an hex number... so I let it go...

3) the matter specification does not define how a controller should handle a device type with multiple clusters that are not mandated or specified as optional.
This is wrong too: FixedLabel is not mandatory or optional with the device types we are considering, but can be added when needed. Custom clusters (e.g. EveHistory) are added by producers to standard device type. I imagine you were still confused by the hex / decimal mistake.

4) in case useful, later in the logs the two lights are flagged as having no unique ID
The unique id for child endpoints is added by the matter.js cause there is no basic info in the child endpoint, so no issue at all

5) the rest is indeed very interesting and I thank you: there is a bug that without analysing the code to answer you, I didn't found before.
Will be updated asap (I need to refactor part of the code there so will not be fast. I will contact you for testing a beta if you like.). Great job!

I just add that we tested with Alexa and it didn't show the LevelControl in a case similar. We will test again till we understand also that other issue that covered yours. I will keep you posted of course.

To analyse the device is easier and faster to use this like I told you (I kept for granted that you already checked the composition of the node):
<img width="1199" alt="image" src="https://github.com/Luligu/matterbridge/assets/132135057/eb0d5030-824a-4c04-a427-6b9770e57df9">

Thanks again for contributing to find the bug.
